File: sdks/dotnet/tests/Croniq.Runner.Sdk.Conformance.Tests/ConformanceRunner.cs
using System.Diagnostics;
using System.Text.Json;

using Croniq.Runner.Sdk;
using Croniq.Runner.Sdk.Configuration;
using Croniq.Runner.Sdk.DependencyInjection;

using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace Croniq.Runner.Sdk.Conformance.Tests;

/// <summary>
/// Orchestrator for a single conformance case: starts a scripted mock
/// server, wires the SDK's DI, drives <see cref="CroniqRunner.RunAsync"/>
/// until expectations are met (or the case deadline elapses), then asserts
/// the recorded HTTP traffic.
/// </summary>
internal static class ConformanceRunner
{
    public static async Task RunAsync(CaseSpec spec, CancellationToken ct = default)
    {
        await using var mock = new MockServerHarness(spec.ServerScript);

        await using var services = BuildServices(spec, mock.BaseUrl);
        var runner = services.GetRequiredService<CroniqRunner>();

        var deadline = TimeSpan.FromMilliseconds(spec.Expectations.DurationMaxMs ?? 5000);

        using var caseCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        caseCts.CancelAfter(deadline);

        // Optional binding directive: cancel the runner partway through.
        // Useful for drain/shutdown cases.
        using var runnerCts = CancellationTokenSource.CreateLinkedTokenSource(caseCts.Token);
        if (spec.ShutdownAfterMs is int after)
        {
            _ = Task.Delay(after, caseCts.Token)
                .ContinueWith(_ => runnerCts.Cancel(), TaskScheduler.Default);
        }

        var runTask = runner.RunAsync(runnerCts.Token);

        // Poll the mock's request log until expectations are satisfied; this
        // lets us exit early once the runner has produced everything we need,
        // instead of always waiting the full duration_max_ms.
        var stopwatch = Stopwatch.StartNew();
        while (!caseCts.IsCancellationRequested)
        {
            if (ExpectationsAreMet(spec.Expectations, mock.RecordedRequests))
            {
                break;
            }
            try
            {
                await Task.Delay(50, caseCts.Token);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }

        runnerCts.Cancel();
        try
        {
            await runTask;
        }
        catch (OperationCanceledException)
        {
            // expected when we cancel the runner
        }
        catch (PollInstanceConflictException)
        {
            // Expected for case 11a (and any future case where the SDK is
            // contractually allowed to bail out fatally). The HTTP-count
            // assertions below still validate correctness: a case that
            // doesn't anticipate this exit would fail on min_count/max_count
            // mismatch.
        }
        stopwatch.Stop();

        var recorded2 = mock.RecordedRequests;
        if (Environment.GetEnvironmentVariable("CRONIQ_CONFORMANCE_DEBUG") == "1")
        {
            Console.WriteLine($"[debug-{spec.Name}] {recorded2.Count} request(s) in {stopwatch.ElapsedMilliseconds}ms:");
            foreach (var r in recorded2) Console.WriteLine($"[debug]   {r.Method} {r.Path}");
        }
        AssertExpectations(spec, recorded2, stopwatch.Elapsed);
    }

    private static ServiceProvider BuildServices(CaseSpec spec, string serverUrl)
    {
        var sc = new ServiceCollection();
        sc.AddLogging(b => b.SetMinimumLevel(LogLevel.Warning));
        sc.AddSingleton<TimeProvider>(TimeProvider.System);

        var builder = sc.AddCroniqRunner(opts =>
        {
            opts.ServerUrl = serverUrl;
            ApplyConfig(opts, spec.RunnerConfig);
        });

        HandlerSentinels.ApplyTo(builder, spec.Handlers);

        // Strip the BackgroundService — we drive RunAsync ourselves so the
        // case can observe the runner's lifecycle deterministically.
        var hostedDescriptor = sc.FirstOrDefault(d => d.ServiceType.FullName == "Microsoft.Extensions.Hosting.IHostedService");
        if (hostedDescriptor is not null)
        {
            sc.Remove(hostedDescriptor);
        }

        return sc.BuildServiceProvider();
    }

    private static void ApplyConfig(CroniqRunnerOptions opts, RunnerConfigSpec cfg)
    {
        if (cfg.RunnerId is not null) opts.RunnerId = cfg.RunnerId;
        if (cfg.RunnerIdPrefix is not null) opts.RunnerIdPrefix = cfg.RunnerIdPrefix;
        foreach (var cap in cfg.Capabilities) opts.Capabilities.Add(cap);
        foreach (var t in cfg.Tags) opts.Tags.Add(t);
        if (cfg.MaxInflight is int m) opts.MaxInflight = m;
        if (cfg.ApiKey is not null) opts.ApiKey = cfg.ApiKey;
        if (cfg.BearerToken is not null) opts.BearerToken = cfg.BearerToken;
        if (cfg.PollTimeoutMs is int pt) opts.PollTimeout = TimeSpan.FromMilliseconds(pt);
        if (cfg.RenewIntervalMs is int ri) opts.RenewInterval = TimeSpan.FromMilliseconds(ri);
        if (cfg.DrainTimeoutMs is int dt) opts.DrainTimeout = TimeSpan.FromMilliseconds(dt);
        if (cfg.PollRetryDelayMs is int prd) opts.PollRetryDelay = TimeSpan.FromMilliseconds(prd);
        if (cfg.CapacityBackoffMs is int cb) opts.CapacityBackoff = TimeSpan.FromMilliseconds(cb);
        if (cfg.MaxConsecutivePollConflicts is int mcpc) opts.MaxConsecutivePollConflicts = mcpc;
    }

    private static bool ExpectationsAreMet(ExpectationsSpec expectations, IReadOnlyList<RecordedRequest> recorded)
    {
        foreach (var ex in expectations.Http)
        {
            // max_count is a "ceiling over a time window" assertion. If
            // we'd let the loop exit as soon as min/exact were satisfied,
            // a runner that violates max only AFTER the early exit would
            // pass trivially. Force the case to wait its full
            // duration_max_ms in that case so the ceiling is observable.
            if (ex.MaxCount.HasValue) return false;

            var matching = recorded.Count(r =>
                r.Method.Equals(ex.Method, StringComparison.OrdinalIgnoreCase) &&
                r.Path == ex.Path);

            if (ex.ExactCount is int exact && matching < exact) return false;
            if (ex.MinCount is int min && matching < min) return false;
            // body_match and header checks happen post-hoc.
        }
        return true;
    }

    private static void AssertExpectations(CaseSpec spec, IReadOnlyList<RecordedRequest> recorded, TimeSpan elapsed)
    {
        // duration_max_ms is enforced as a wall-clock deadline by
        // caseCts.CancelAfter(deadline). An additional post-hoc
        // ShouldBeLessThan check on `elapsed` is redundant AND wrong for
        // cases that legitimately run the full window (max_count
        // assertions, see ExpectationsAreMet): the stopwatch always reads
        // a few ms over the deadline because cancellation propagation
        // isn't instantaneous. We rely on the cancellation enforcement
        // and skip the redundant assertion.
        _ = elapsed;

        foreach (var ex in spec.Expectations.Http)
        {
            var matches = recorded
                .Where(r => r.Method.Equals(ex.Method, StringComparison.OrdinalIgnoreCase) && r.Path == ex.Path)
                .ToList();

            if (ex.ExactCount is int exact)
            {
                matches.Count.ShouldBe(exact, $"{ex.Method} {ex.Path}: expected exact_count={exact}");
            }
            if (ex.MinCount is int min)
            {
                matches.Count.ShouldBeGreaterThanOrEqualTo(min, $"{ex.Method} {ex.Path}: expected min_count={min}");
            }
            if (ex.MaxCount is int xmax)
            {
                matches.Count.ShouldBeLessThanOrEqualTo(xmax, $"{ex.Method} {ex.Path}: expected max_count={xmax}");
            }

            if (ex.Headers.Count > 0 && matches.Count > 0)
            {
                // The first matching request must carry all expected headers.
                var first = matches[0];
                foreach (var (name, expected) in ex.Headers)
                {
                    first.Headers.ContainsKey(name).ShouldBeTrue($"{ex.Method} {ex.Path}: missing header '{name}'");
                    var actual = first.Headers[name];
                    if (expected == "*")
                    {
                        actual.ShouldNotBeNullOrEmpty($"{ex.Method} {ex.Path}: header '{name}' was empty");
                    }
                    else
                    {
                        actual.ShouldBe(expected, $"{ex.Method} {ex.Path}: header '{name}' mismatch");
                    }
                }
            }

            if (ex.BodyMatch is not null && matches.Count > 0)
            {
                // First matching request must satisfy the body subset match.
                using var doc = JsonDocument.Parse(matches[0].Body);
                var err = BodyMatcher.Match(ex.BodyMatch, doc.RootElement);
                err.ShouldBeNull($"{ex.Method} {ex.Path}: body mismatch — {err}");
            }
        }
    }
}

File: sdks/java/conformance-tests/src/test/java/io/croniq/runner/conformance/ConformanceRunner.java
package io.croniq.runner.conformance;

import static org.assertj.core.api.Assertions.fail;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.croniq.runner.CroniqAuthFailedException;
import io.croniq.runner.CroniqOwnershipDeniedException;
import io.croniq.runner.CroniqPollInstanceConflictException;
import io.croniq.runner.CroniqRunner;
import io.croniq.runner.config.CroniqRunnerOptions;
import java.time.Duration;
import java.util.List;
import org.awaitility.Awaitility;

/**
 * Drives one conformance case end-to-end: stand up the mock server, build
 * a {@link CroniqRunner} with the case's config, run it on a worker thread,
 * poll for expectations, then assert the recorded request stream.
 *
 * <p>Lifetime: each instance is single-use. Call {@link #run(CaseSpec)} once
 * per case.
 */
final class ConformanceRunner {

    private static final ObjectMapper JSON = new ObjectMapper();

    void run(CaseSpec spec) throws Exception {
        try (MockServerHarness server = new MockServerHarness(spec.serverScript())) {
            server.start();
            CroniqRunnerOptions options = buildOptions(spec.runnerConfig(), server.baseUrl());

            CroniqRunner.Builder builder = CroniqRunner.builder().options(options);
            HandlerSentinels.applyTo(builder, spec.handlers());

            CroniqRunner runner = builder.build();
            Thread loop = Thread.ofVirtual().name("conformance-runner-loop").start(() -> {
                try {
                    runner.run();
                } catch (InterruptedException ignored) {
                    // expected on close()
                } catch (CroniqOwnershipDeniedException
                        | CroniqPollInstanceConflictException
                        | CroniqAuthFailedException ignored) {
                    // Expected for cases 15, 16 and 17: a 403 on poll is
                    // permanent, a streak of 409s exhausts the conflict
                    // ceiling, and a streak of 401s exhausts the auth ceiling
                    // — the SDK is contractually required to stop in all
                    // three. The HTTP-count assertions are what prove it
                    // actually did; a case that doesn't anticipate this exit
                    // still fails on min_count/max_count.
                }
            });
            try {
                // Optional mid-run shutdown directive (drain cases — PR-3).
                if (spec.shutdownAfterMs() != null) {
                    Thread.sleep(spec.shutdownAfterMs());
                    runner.close();
                }

                // Wait for the expectations to be satisfiable or the case's
                // duration cap, whichever comes first. The wait is the
                // "settling" period — there's no early-exit ack from the
                // SDK so we just let the loop run until the deadline.
                int durationMaxMs = spec.expectations().durationMaxMs() == null
                        ? 5_000
                        : spec.expectations().durationMaxMs();
                if (hasMaxCount(spec)) {
                    // A max_count is a "ceiling over a time window" assertion.
                    // Exiting as soon as the lower bounds are met would let a
                    // runner that violates the ceiling right after our exit
                    // pass trivially, so burn the full window. Mirrors the
                    // .NET / Go / Python / TypeScript bindings.
                    Thread.sleep(durationMaxMs);
                } else {
                    Awaitility.await()
                            .atMost(Duration.ofMillis(durationMaxMs))
                            .pollInterval(Duration.ofMillis(50))
                            .until(() -> expectationsLikelyMet(spec, server.recorded()));
                }
            } finally {
                runner.close();
                loop.join(Duration.ofSeconds(5));
            }

            assertExpectations(spec, server.recorded());
        }
    }

    private static CroniqRunnerOptions buildOptions(CaseSpec.RunnerConfig rc, String serverUrl) {
        var b = CroniqRunnerOptions.builder().serverUrl(serverUrl);
        if (rc == null) {
            return b.build();
        }
        if (rc.runnerId() != null) {
            b.runnerId(rc.runnerId());
        }
        if (rc.runnerIdPrefix() != null) {
            b.runnerIdPrefix(rc.runnerIdPrefix());
        }
        if (rc.capabilities() != null) {
            b.capabilities(rc.capabilities());
        }
        if (rc.tags() != null) {
            b.tags(rc.tags());
        }
        if (rc.maxInflight() != null) {
            b.maxInflight(rc.maxInflight());
        }
        if (rc.apiKey() != null) {
            b.apiKey(rc.apiKey());
        }
        if (rc.bearerToken() != null) {
            b.bearerToken(rc.bearerToken());
        }
        if (rc.pollTimeoutMs() != null) {
            b.pollTimeout(Duration.ofMillis(rc.pollTimeoutMs()));
        }
        if (rc.renewIntervalMs() != null) {
            b.renewInterval(Duration.ofMillis(rc.renewIntervalMs()));
        }
        if (rc.drainTimeoutMs() != null) {
            b.drainTimeout(Duration.ofMillis(rc.drainTimeoutMs()));
        }
        if (rc.pollRetryDelayMs() != null) {
            b.pollRetryDelay(Duration.ofMillis(rc.pollRetryDelayMs()));
        }
        if (rc.capacityBackoffMs() != null) {
            b.capacityBackoff(Duration.ofMillis(rc.capacityBackoffMs()));
        }
        if (rc.maxConsecutivePollConflicts() != null) {
            b.maxConsecutivePollConflicts(rc.maxConsecutivePollConflicts());
        }
        if (rc.maxConsecutiveAuthFailures() != null) {
            b.maxConsecutiveAuthFailures(rc.maxConsecutiveAuthFailures());
        }
        return b.build();
    }

    /** True when any expectation carries a {@code max_count} ceiling. */
    private static boolean hasMaxCount(CaseSpec spec) {
        if (spec.expectations() == null || spec.expectations().http() == null) {
            return false;
        }
        return spec.expectations().http().stream().anyMatch(e -> e.maxCount() != null);
    }

    /**
     * Returns true once every expectation's {@code min_count} / {@code exact_count}
     * lower bound is satisfied. Only consulted for cases without a
     * {@code max_count} — those burn the full window instead, see
     * {@link #hasMaxCount(CaseSpec)}.
     */
    private static boolean expectationsLikelyMet(CaseSpec spec, List<MockServerHarness.RecordedRequest> recorded) {
        if (spec.expectations() == null || spec.expectations().http() == null) {
            return true;
        }
        for (var e : spec.expectations().http()) {
            long n = recorded.stream()
                    .filter(r -> r.matches(e.method(), e.path()))
                    .count();
            if (e.exactCount() != null && n < e.exactCount()) {
                return false;
            }
            if (e.minCount() != null && n < e.minCount()) {
                return false;
            }
        }
        return true;
    }

    private static void assertExpectations(CaseSpec spec, List<MockServerHarness.RecordedRequest> recorded) {
        if (spec.expectations() == null || spec.expectations().http() == null) {
            return;
        }
        for (var e : spec.expectations().http()) {
            var matches = recorded.stream()
                    .filter(r -> r.matches(e.method(), e.path()))
                    .toList();
            int n = matches.size();
            if (e.exactCount() != null && n != e.exactCount()) {
                fail("Expected %s %s exact_count=%d, got %d. Recorded: %s"
                        .formatted(e.method(), e.path(), e.exactCount(), n, summary(recorded)));
            }
            if (e.minCount() != null && n < e.minCount()) {
                fail("Expected %s %s min_count=%d, got %d".formatted(e.method(), e.path(), e.minCount(), n));
            }
            if (e.maxCount() != null && n > e.maxCount()) {
                fail("Expected %s %s max_count=%d, got %d".formatted(e.method(), e.path(), e.maxCount(), n));
            }
            if (e.headers() != null && !e.headers().isEmpty()) {
                if (matches.isEmpty()) {
                    fail("Headers expected on %s %s but no requests recorded".formatted(e.method(), e.path()));
                }
                var first = matches.get(0);
                for (var h : e.headers().entrySet()) {
                    String actual = first.headers().get(h.getKey().toLowerCase());
                    if (actual == null) {
                        fail("Missing header '%s' on %s %s. Headers seen: %s"
                                .formatted(
                                        h.getKey(),
                                        e.method(),
                                        e.path(),
                                        first.headers().keySet()));
                    }
                    if ("*".equals(h.getValue())) {
                        if (actual.isEmpty()) {
                            fail("Header '%s' expected non-empty (*) but was empty".formatted(h.getKey()));
                        }
                    } else if (!h.getValue().equals(actual)) {
                        fail("Header '%s' expected '%s' but was '%s'".formatted(h.getKey(), h.getValue(), actual));
                    }
                }
            }
            if (e.bodyMatch() != null) {
                if (matches.isEmpty()) {
                    fail("body_match expected on %s %s but no requests recorded".formatted(e.method(), e.path()));
                }
                var first = matches.get(0);
                JsonNode actualBody;
                try {
                    actualBody = first.body().isEmpty() ? JSON.nullNode() : JSON.readTree(first.body());
                } catch (Exception ex) {
                    fail("Could not parse JSON body for %s %s: %s".formatted(e.method(), e.path(), first.body()));
                    return;
                }
                String err = BodyMatcher.match(e.bodyMatch(), actualBody);
                if (err != null) {
                    fail("body_match failed on %s %s: %s. Actual: %s"
                            .formatted(e.method(), e.path(), err, first.body()));
                }
            }
        }
    }

    private static String summary(List<MockServerHarness.RecordedRequest> recorded) {
        StringBuilder sb = new StringBuilder().append('[');
        for (var r : recorded) {
            sb.append(r.method()).append(' ').append(r.path()).append(", ");
        }
        if (sb.length() > 1) {
            sb.setLength(sb.length() - 2);
        }
        return sb.append(']').toString();
    }
}
